File: change_golden_test.go
// Copyright 2026 Jasen Minton
//
// SPDX-License-Identifier: Apache-2.0
package main

import (
	"encoding/json"
	"math"
	"os"
	"reflect"
	"testing"
)

type changeGoldenCorpus struct {
	Cases []changeGoldenCase `json:"cases"`
}

type changeGoldenCase struct {
	Name      string            `json:"name"`
	Reference changeGoldenInput `json:"reference"`
	Selected  changeGoldenInput `json:"selected"`
	Expected  changeGoldenWant  `json:"expected"`
}

type changeGoldenWant struct {
	Score      float64                 `json:"score"`
	Primary    string                  `json:"primary"`
	Components []changeGoldenComponent `json:"components"`
}

type changeGoldenComponent struct {
	Key   string  `json:"key"`
	Score float64 `json:"score"`
}

type changeGoldenInput struct {
	Lines     int                 `json:"lines"`
	Bytes     int                 `json:"bytes"`
	Errors    int                 `json:"errors"`
	Marked    int                 `json:"marked"`
	B16       int                 `json:"b16"`
	WordPeaks []changeGoldenEntry `json:"word_peaks"`
	RefPeaks  []changeGoldenEntry `json:"ref_peaks"`
	IPPeaks   []changeGoldenEntry `json:"ip_peaks"`
	WordWave  []changeGoldenEntry `json:"word_wave"`
}

type changeGoldenEntry struct {
	Key   string `json:"key"`
	Count int    `json:"count"`
}

func TestChangeGoldenCorpus(t *testing.T) {
	data, err := os.ReadFile("testdata/change_golden.json")
	if err != nil {
		t.Fatal(err)
	}
	var corpus changeGoldenCorpus
	if err := json.Unmarshal(data, &corpus); err != nil {
		t.Fatal(err)
	}

	for _, test := range corpus.Cases {
		t.Run(test.Name, func(t *testing.T) {
			result := compareChangeShapes(
				goldenChangeShape(test.Reference),
				goldenChangeShape(test.Selected),
			)
			if math.Abs(result.score-test.Expected.Score) > 1e-9 {
				t.Fatalf("score = %.15f, want %.15f", result.score, test.Expected.Score)
			}
			if result.components[0].key != test.Expected.Primary {
				t.Fatalf("primary = %q, want %q", result.components[0].key, test.Expected.Primary)
			}
			gotKeys := make([]string, len(result.components))
			for i, component := range result.components {
				gotKeys[i] = component.key
			}
			wantKeys := make([]string, len(test.Expected.Components))
			for i, component := range test.Expected.Components {
				wantKeys[i] = component.Key
			}
			if !reflect.DeepEqual(gotKeys, wantKeys) {
				t.Fatalf("component order = %v, want %v", gotKeys, wantKeys)
			}
			for i, component := range result.components {
				if math.Abs(component.score-test.Expected.Components[i].Score) > 1e-9 {
					t.Fatalf("component %q = %.15f, want %.15f",
						component.key, component.score, test.Expected.Components[i].Score)
				}
			}
		})
	}
}

func goldenChangeShape(input changeGoldenInput) changeShape {
	shape := changeShape{
		lines:     float64(input.Lines),
		bytes:     float64(input.Bytes),
		errors:    float64(input.Errors),
		wordPeaks: goldenDistribution(input.WordPeaks),
		refPeaks:  goldenDistribution(input.RefPeaks),
		ipPeaks:   goldenDistribution(input.IPPeaks),
		wordWave:  goldenDistribution(input.WordWave),
	}
	if input.Lines > 0 {
		shape.averageBytes = changeOptionalValue{value: float64(input.Bytes) / float64(input.Lines), available: true}
		shape.marked = changeOptionalValue{value: float64(input.Marked) * 100 / float64(input.Lines), available: true}
		shape.b16 = changeOptionalValue{value: float64(input.B16) * 100 / float64(input.Lines), available: true}
	}
	return shape
}

func goldenDistribution(entries []changeGoldenEntry) []changeDistributionEntry {
	result := make([]changeDistributionEntry, len(entries))
	for i, entry := range entries {
		result[i] = changeDistributionEntry{key: entry.Key, count: entry.Count}
	}
	return result
}

File: cmd/pattyView/src/domain/changeGolden.test.ts
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import { buildChangeSeries, type ChangeComponentKey } from './changeAnalysis'
import { parseJsonlText } from './jsonl'

interface GoldenEntry {
  key: string
  count: number
}

interface GoldenInput {
  lines: number
  bytes: number
  errors: number
  marked: number
  b16: number
  word_peaks: GoldenEntry[]
  ref_peaks: GoldenEntry[]
  ip_peaks: GoldenEntry[]
  word_wave: GoldenEntry[]
}

interface GoldenComponent {
  key: string
  score: number
}

interface GoldenCase {
  name: string
  reference: GoldenInput
  selected: GoldenInput
  expected: {
    score: number
    primary: string
    components: GoldenComponent[]
  }
}

const corpus = JSON.parse(readFileSync(
  resolve(process.cwd(), '../../testdata/change_golden.json'),
  'utf8',
)) as { cases: GoldenCase[] }

const canonicalComponent: Record<ChangeComponentKey, string> = {
  lines: 'lines',
  bytes: 'bytes',
  averageBytes: 'avg bytes/line',
  errors: 'errs',
  marked: 'marked',
  b16: 'b16',
  peakBalance: 'peak balance',
  wordWave: 'word wave',
}

describe('shared Change golden corpus', () => {
  for (const golden of corpus.cases) {
    it(golden.name, () => {
      const records = parseJsonlText([
        goldenInterval(0, golden.reference),
        goldenInterval(1, golden.selected),
      ].map((record) => JSON.stringify(record)).join('\n')).records
      const point = buildChangeSeries(records)[1]

      expect(Math.abs((point.score ?? -1) - golden.expected.score)).toBeLessThanOrEqual(1e-9)
      expect(point.primary === null ? null : canonicalComponent[point.primary]).toBe(golden.expected.primary)
      expect(point.components.map((component) => canonicalComponent[component.key])).toEqual(
        golden.expected.components.map((component) => component.key),
      )
      point.components.forEach((component, index) => {
        expect(Math.abs(component.score - golden.expected.components[index].score)).toBeLessThanOrEqual(1e-9)
      })
    })
  }
})

function goldenInterval(number: number, input: GoldenInput) {
  const collection = (entries: GoldenEntry[], isPeak: boolean) => entries.map((entry) => ({
    key: entry.key,
    count: entry.count,
    is_peak: isPeak,
  }))
  const stream = (name: string, top: GoldenEntry[], peaks: GoldenEntry[]) => ({
    name,
    top: collection(top, false),
    peaks: collection(peaks, true),
  })

  return {
    schema_version: 4,
    event_type: 'interval',
    session_id: 'change-golden',
    interval: number,
    interval_lines: input.lines,
    matchers: [
      {
        name: 'lines',
        interval_count: input.lines,
        top_keys: [
          { key: 'marked', count: input.marked },
          { key: ' b16', count: input.b16 },
        ],
      },
      { name: 'bytes', interval_count: input.bytes },
      { name: 'errs', interval_count: input.errors },
    ],
    interesting: [
      stream('words', input.word_wave, input.word_peaks),
      stream('refs', [], input.ref_peaks),
      stream('ips', [], input.ip_peaks),
    ],
  }
}
